src/Highlighter.js

```diff
@@ -18,190 +18,153 @@ const _generateComponentKey = (usedKeys, id) => {
+    const { offsetLeft, offsetTop } = caretElement
 
-    if (this.state.left === offsetLeft && this.state.top === offsetTop) {
+    if (position.left === offsetLeft && position.top === offsetTop) {
       return
     }
 
-    const newPosition = {
-      left: offsetLeft,
-      top: offsetTop,
-    }
-    this.setState(newPosition)
+    const newPosition = { left: offsetLeft, top: offsetTop }
+    setPosition(newPosition)
 
-    this.props.onCaretPositionChange(newPosition)
+    onCaretPositionChange(newPosition)
   }
 
-  render() {
-    const {
-      selectionStart,
-      selectionEnd,
-      value,
-      style,
-      children,
-      containerRef,
-    } = this.props
-    const config = readConfigFromChildren(children)
-
-    // If there's a caret (i.e. no range selection), map the caret position into the marked up value
-    let caretPositionInMarkup
-    if (selectionStart === selectionEnd) {
-      caretPositionInMarkup = mapPlainTextIndex(
-        value,
-        config,
-        selectionStart,
-        'START'
-      )
-    }
+  const config = readConfigFromChildren(children)
+  let caretPositionInMarkup
 
-    const resultComponents = []
-    const componentKeys = {}
-
-    // start by appending directly to the resultComponents
-    let components = resultComponents
-    let substringComponentKey = 0
-
-    const textIteratee = (substr, index, indexInPlainText) => {
-      // check whether the caret element has to be inserted inside the current plain substring
-      if (
-        isNumber(caretPositionInMarkup) &&
-        caretPositionInMarkup >= index &&
-        caretPositionInMarkup <= index + substr.length
-      ) {
-        // if yes, split substr at the caret position and insert the caret component
-        const splitIndex = caretPositionInMarkup - index
-        components.push(
-          this.renderSubstring(
-            substr.substring(0, splitIndex),
-            substringComponentKey
-          )
-        )
-
-        // add all following substrings and mention components as children of the caret component
-        components = [
-          this.renderSubstring(
-            substr.substring(splitIndex),
-            substringComponentKey
-          ),
-        ]
-      } else {
-        // otherwise just push the plain text substring
-        components.push(this.renderSubstring(substr, substringComponentKey))
-      }
-
-      substringComponentKey++
-    }
+  if (selectionEnd === selectionStart) {
+    caretPositionInMarkup = mapPlainTextIndex(
+      value,
+      config,
+      selectionStart,
+      'START'
+    )
+  }
 
-    const mentionIteratee = (
-      markup,
-      index,
-      indexInPlainText,
-      id,
-      display,
-      mentionChildIndex,
-      lastMentionEndIndex
-    ) => {
-      // generate a component key based on the id
-      const key = _generateComponentKey(componentKeys, id)
+  const resultComponents = []
+  const componentKeys = {}
+  let components = resultComponents
```

Comment:
Missing original comment: `// start by appending directly to the resultComponents`